File: host-mat/demos/basic/basic.m
%% DEMO: Basic Tx/Rx with the Sivers 60 GHz, 1 channel SDR

%% Packages
% Add the folder containing +mmwsdr to the MATLAB path.
addpath('../../');

%% Parameters
%
%                              | Control    | Data 1     | Data 2     |
% sdr2-in1.sb1.cosmos-lab.org: | 10.113.6.3 | 10.114.6.3 | 10.115.6.3 |
% sdr2-in2.sb1.cosmos-lab.org: | 10.113.6.4 | 10.114.6.4 | 10.115.6.4 |
%
ip = "10.1.1.43";	% IP Address
isDebug = true;		% print debug messages

%% Create a SDR
sdr0 = mmwsdr.sdr.Sivers60GHz('ip', ip, 'isDebug', isDebug);

% Configure the RFSoC
sdr0.fpga.configure('../../config/rfsoc.cfg');

%% Send data
nFFT = 1024;            % number of FFT points
txPower = 10000*1;      % transmit
scMin = -250;
scMax = 250;
constellation = [1+1j 1-1j -1+1j -1-1j];

txfd = zeros(nFFT, sdr0.nch);
txfd(nFFT/2 + 1 + (scMin:scMax),:) = ...
    constellation(randi(4,length(scMin:scMax),sdr0.nch));
txfd = fftshift(txfd);
txtd = ifft(txfd);

% Normalize the energy of the tx array and scale with txPower.
txtd = txPower*txtd./max(abs(txtd));

sdr0.send(txtd);

%% Receive data

% To read data from the ADCs we use the `recv` method of the FullyDigital
% sdr class. This method has 3 arguments.
% * nsamp: number of continuous samples to read
% * nskip: number of samples to skip
% * nbatch: number of batches

nFFT = 1024;	% num of FFT points
nskip = 1024*3;	% skip ADC data for 1024 cc
nbatch = 200;	% num of batches

rxtd = sdr0.recv(nFFT, nskip, nbatch);

%% Plot the received data
scs = linspace(-nFFT/2, nFFT/2-1, nFFT);
for ibatch=1:10
    % Plot the frequency-domain signal
    f = figure(2);
    plot(scs, mag2db(abs(fftshift(fft(rxtd(:,ibatch))))));
    axis tight; grid on; grid minor;
    ylabel('Magnitude [dB]', 'interpreter', 'latex', 'fontsize', 12);
    xlabel('Subcarrier Index', 'interpreter', 'latex', 'fontsize', 12);
end

%% Close the TCP connection to the SDR
clear sdr0

File: host/demos/basic/basic.py
"""

"""

# Import Libraries
import os
import sys
import argparse
import numpy as np
import matplotlib

matplotlib.use('TkAgg')
from matplotlib import pyplot as plt

path = os.path.abspath('../../')
if not path in sys.path:
    sys.path.append(path)
import mmwsdr

# Parameters
nfft = 1024  # num of continuous samples per batch
nskip = 1024 * 5  # num of samples to skip between batches
nbatch = 10  # num of batches
isdebug = True  # print debug messages
sc = 400  # subcarrier index
tx_pwr = 10000  # transmit power


def main():
    """

    :return:
    :rtype:
    """
    parser = argparse.ArgumentParser()
    parser.add_argument("--freq", type=float, default=60.48e9, help="receiver carrier frequency in Hz (i.e., 60.48e9)")
    parser.add_argument("--node", type=str, default='sdr2-in1', help="cosmos-sb1 node name (i.e., sdr2-in1)")
    parser.add_argument("--mode", type=str, default='rx', help="sdr mode (i.e., rx)")
    args = parser.parse_args()

    # Create an SDR object and the XY table
    if args.node == 'sdr2-in1':
        sdr0 = mmwsdr.sdr.Sivers60GHz(ip='10.113.6.3', freq=args.freq, unit_name='SN0240', isdebug=isdebug)
        xytable0 = mmwsdr.utils.XYTable('xytable1')

        # Move the SDR to the lower-right corner
        xytable0.move(x=0, y=0, angle=0)
    elif args.node == 'sdr2-in2':
        sdr0 = mmwsdr.sdr.Sivers60GHz(ip='10.113.6.4', freq=args.freq, unit_name='SN0243', isdebug=isdebug)
        xytable0 = mmwsdr.utils.XYTable('xytable2')

        # Move the SDR to the lower-left conrner
        xytable0.move(x=1300, y=0, angle=0)
    else:
        raise ValueError("COSMOS node can be either 'sdr2-in1' or 'sdr2-in2'")

    # Configure the RFSoC
    sdr0.fpga.configure('../../config/rfsoc.cfg')

    # Main loop
    while (1):
        if args.mode == 'tx':
            # Make sure that the nodes are not transmitting
            sdr0.send(np.zeros((nfft,), dtype='int16'))

            # Create a signal in frequency domain
            txfd = np.zeros((nfft,), dtype='int16')
            txfd[(nfft >> 1) + sc] = 1
            txfd = np.fft.fftshift(txfd, axes=0)

            # Then, convert it to time domain
            txtd = np.fft.ifft(txfd, axis=0)

            # Set the tx power
            txtd = txtd / np.mean(np.abs(txfd)) * tx_pwr

            # Transmit data
            sdr0.send(txtd)
        elif args.mode == 'rx':
            # Receive data
            rxtd = sdr0.recv(nfft, nskip, nbatch)

            rxfd = np.fft.fft(rxtd, axis=1)
            rxfd = np.fft.fftshift(rxfd, axes=1)
            f = np.linspace(-nfft / 2, nfft / 2 - 1, nfft)
            plt.plot(f, abs(np.mean(rxfd, axis=0)), '-')
            plt.xlabel('Subcarrier index')
            plt.xlabel('Magnitude')
            plt.tight_layout()
            plt.grid()
            plt.show()
        else:
            raise ValueError("SDR mode can be either 'tx' or 'rx'")

        if sys.version_info[0] == 2:
            ans = raw_input("Enter 'q' to exit or\n press enter to continue ")
        else:
            ans = input("Enter 'q' to exit or\n press enter to continue ")

        if ans == 'q':
            break
    # Close the TPC connections
    del sdr0


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass
